altair/utils/core.py

```diff
@@ -200,7 +201,12 @@ def infer_dtype(value):
 ]
 
 
-def infer_vegalite_type(data):
+_InferredVegaLiteType = Literal["ordinal", "nominal", "quantitative", "temporal"]
+
+
+def infer_vegalite_type(
+    data: Union[np.ndarray, pd.Series]
```

Comment:
Another question, here you mention: `data: Union[np.ndarray, pd.Series]`. When I look to where this is used: https://github.com/altair-viz/altair/blob/master/altair/vegalite/v5/api.py#L2411. I cannot understand why we need to be restrictive on the input types from numpy and pandas. I really like types like `DataFrameLike` and `SupportsGeoInterface` for input data features.

Reply:
I tried to define a type hint named `ArrayLike1D` something that can serve for one-dimensional data of various forms, but not sure if this is possible yet (eg. https://github.com/python/mypy/issues/12280).

Otherwise, maybe something like they define as type hint for the `values` of a polars series? There they define an `ArrayLike` type as such (see [here](https://github.com/pola-rs/polars/blob/main/py-polars/polars/series/series.py#L139-L147):
```python
ArrayLike = Union[
    Sequence[Any],
    "Series",
    "pa.Array",
    "pa.ChunkedArray",
    "np.ndarray",
    "pd.Series",
    "pd.DatetimeIndex",
]
```
And in the docstring for values of the series they mention:
```
values : ArrayLike, default None
    One-dimensional data in various forms. Supported are: Sequence, Series,
    pyarrow Array, and numpy ndarray.
```
I don't think this type-hint is strictly 1D as the docstring suggest, but it can serve as a reference. 
Somehow wished it would be possible to include `range` as well (ref: https://github.com/altair-viz/altair/issues/2877).

Reply:
You're right, that's too restrictive. I used the type hints which were already present in the docstring but seems like this function supports anything that the pandas function `infer_type` can handle. According to [their own type hints](https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/lib.pyx#L1403) this is all `object` so basically everything. I'd suggest that we type hint it the same to be consistent with pandas. Implemented in e974bc96d7ba07dc09f0a1d866c00c30c8b34c07

Btw, the function `infer_vegalite_type` is used only [here](https://github.com/altair-viz/altair/blob/master/altair/utils/core.py#L576). Your link points to the usage of `infer_encoding_types`.

Reply:
That is flexible enough, thanks for the change.